design/secure-communication-design.md

```diff
@@ -0,0 +1,88 @@
+
+## Non-goals
+
+- Support for dynamic credentials
+- Support for use of Secret Discovery Service for credential management
+
+## Background
+
+Currently, Contour and Envoy containers are co-located within a pod.
+Current security model is relying on the fact that Contour exposes the management
+server on loopback interface which prevents any unauthorized access from outside
+the pod.
+
+With the need to move away from the [current deployment pattern of co-locating
+Contour and Envoy][0], it is necessary to secure the communication channel between
+the containers.
+
+## High level design
+
+- Use mutual TLS to secure communication channel between Contour and Envoy
+- Bootstrap phase of Contour will generate necessary TLS certificates
+
+## Detailed design
+
+### Default options for secure communication
+
+By default, mutual TLS is enabled to secure the communication between Contour and Envoy
+containers. The default `contour bootstrap` will generate the bootstrap configuration that
+enables secure communication between the containers. If `contour bootstrap` is passed with
+the parameter `--disable-tls`, the bootstrap configuration generated will make use of insecure
+communication between the containers.
+
+### Bootstrapping process
+
+This bootstrap phase is independent of `contour bootstrap` and uses `contour bootstrap-tls`.
+`contour bootstrap-tls` is run as an init container to bootstrap the TLS configuration.
+The bootstrap phase will create three pairs of the TLS certificates. The details about
+each pair of certificates are mentioned below:
+1. The first pair is used by Contour to serve HTTPS.
+2. The second pair is used by Envoy to authenticate with Contour.
+3. The last pair serves as the CA certificate which signs the above certificates.
+
+Each of these TLS certificates will be stored as Kubernetes secrets in namespace `heptio-contour`. The
+secrets will be named as `contour-tls`, `envoy-tls` and `contour-ca`, respectively. This ensures that
+even when the `contour` and `envoy` containers are not part of the same pod, the certificates are
+available for consumption by each of those.
+
+Note: In order for Contour to create the Kubernetes secrets in the `heptio-contour` namespace, the
+existing RBAC needs to be enhanced to grant permissions to `create` secrets in the namespace.
+
+### Setting up secure communication channel
+
+1. Contour container will pick the certificates from the Kubernetes secret `contour-tls` which is
+mounted at `/etc/tls/contour` and pick `contour-ca` mounted at `/etc/tls/ca` to start listening for
+HTTPS traffic.
+2. The bootstrap configuration yaml of Envoy will have [`tls_context`][1] added as part of `cluster` object.
+Envoy container will pick the certificates from the Kubernetes secret `envoy-tls` which is mounted at
+`/etc/tls/envoy` and pick `contour-ca` mounted at `/etc/tls/ca` to connect to Contour as configured by the
+bootstrap configuration file.
+
+### Rotation and invalidation of credentials
```

Comment:
If we regen new certs, how do we roll that change out to the Envoy instances? In the deployment model of a colocated Contour with Envoy, this is easy, but when you split apart Envoy from Contour, you'll have a separate daemonset to roll the Envoy pods.

Reply:
Yes, we will have to document how to handle the case where Contour and Envoy are not colocated. As per the initial discussion I had with @davecheney on #862, we agreed that we can implement and address this as part of #881 where it will be required to maybe have SDS in place (#898).

Reply:
I don't think SDS would solve this issue 100%. I think there's still a timing issue that needs a blocker somehow. Please see comment below on mounting secret from init container. 

Also, we already have a split deployment example as well as Gimbal uses this model by default.

Reply:
@stevesloka you are right that SDS will not solve this issue. I did a bit of reading and this is what I found
> The connection between Envoy proxy and SDS server has to be secure. One option is to run the SDS server on the same host and use Unix Domain Socket for the connection. Otherwise it requires mTLS between the proxy and SDS server. In this case, the client certificates for the SDS connection must be statically configured.

Reference: https://www.envoyproxy.io/docs/envoy/v1.10.0/configuration/secret